library/Booster/Syntax/Json/Internalise.hs

```diff
@@ -269,41 +311,111 @@ internaliseTerm ::
     Except PatternError Internal.Term
 internaliseTerm = internaliseTermRaw IsNotQQ
 
--- Throws errors when a term is encountered. The 'And' case
--- is analysed before, this function produces an 'AndPredicate'.
-internalisePredicate ::
+{- | Internalises an And-ed set of predicates, classifying them into
+  BoolPred, CeilPred, SubstitutionPred, and UnsupportedPred.
+  The substitution (as a whole) is analysed after internalisation, to
+  ensure nothing is circular or ambiguous.
+
+  Throws a PatternError when any of the top-level patterns is a term
+  or an And with a term (must be analysed beforehand).
+-}
+internalisePredicates ::
+    Flag "alias" ->
+    Flag "subsorts" ->
+    Maybe [Syntax.Id] ->
+    KoreDefinition ->
+    [Syntax.KorePattern] ->
+    Except PatternError InternalisedPredicates
+internalisePredicates allowAlias checkSubsorts sortVars definition ps = do
+    internalised <-
+        concatMapM (internalisePred allowAlias checkSubsorts sortVars definition) $
+            concatMap explodeAnd ps
+
+    let (substitution, moreEquations) =
+            mkSubstitution [s | s@SubstitutionPred{} <- internalised]
+
+    pure
+        InternalisedPredicates
+            { boolPredicates = Set.fromList $ [p | BoolPred p <- internalised] <> moreEquations
+            , ceilPredicates = Set.fromList $ [p | CeilPred p <- internalised]
+            , substitution
+            , unsupported = [p | UnsupportedPred p <- internalised]
+            }
+  where
+    mkSubstitution ::
+        [InternalisedPredicate] -> (Map Internal.Variable Internal.Term, [Internal.Predicate])
+    mkSubstitution initialSubst =
```

Comment:
This code is quite cool! Perhaps worth factoring-out the `mkSubstitution` into a top-level function and unit-test it in isolation?

Reply:
I moved this to top-level but the unit tests I wrote are still working on the `internalisePredicates` level.